internal/util/inventoryconverter.go

```diff
@@ -0,0 +1,138 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &rivets.Server{
+		BIOSCfg:  biosCfg,
+		ID:       serverID,
+		Facility: facility,
+		// Name:
+		// BMCAddress:
+		// BMCUser:
+		// BMCPassword: device.BMC.
+		Vendor:     device.Vendor,
+		Model:      device.Model,
+		Serial:     device.Serial,
+		Status:     device.Status.State,
+		Components: components,
+	}, nil
+}
+
+func (ic *InventoryConverter) newComponent(slug, cvendor, cmodel, cserial, cproduct string, attrs *rivets.ComponentAttributes, status *common.Status, firmware *common.Firmware) *rivets.Component {
+	slug = strings.ToLower(slug)
+	_, exists := ic.componentSlug[slug]
+	if !exists {
+		return nil
+	}
+
+	if strings.TrimSpace(cmodel) == "" && strings.TrimSpace(cproduct) != "" {
+		cmodel = cproduct
+	}
+	return &rivets.Component{
+		ID:         ic.componentSlug[slug].ID,
+		Firmware:   firmware,
+		Status:     status,
+		Attributes: attrs,
+		Name:       ic.componentSlug[slug].Name,
+		Vendor:     cvendor,
+		Model:      cmodel,
+		Serial:     cserial,
+	}
+}
+
+func (ic *InventoryConverter) getComponentSlice(device *common.Device) ([]*rivets.Component, error) {
+	if len(ic.componentSlug) == 0 {
+		return nil, fmt.Errorf("omponent slugs lookup map empty")
+	}
+	components := []*rivets.Component{}
+
+	// bios
+	devBIOS := device.BIOS
+	component := ic.newComponent(common.SlugBIOS, devBIOS.Vendor, devBIOS.Model, devBIOS.Serial, devBIOS.ProductName,
+		&rivets.ComponentAttributes{
+			Description:   devBIOS.Description,
+			ProductName:   devBIOS.ProductName,
+			SizeBytes:     devBIOS.SizeBytes,
+			CapacityBytes: devBIOS.CapacityBytes,
+			Oem:           devBIOS.Oem,
+			Metadata:      devBIOS.Metadata,
+			Capabilities:  devBIOS.Capabilities,
+		},
+		devBIOS.Status,
+		devBIOS.Firmware,
+	)
+	if component != nil {
+		components = append(components, component)
+	}
+
+	// bmc
+	devBMC := device.BMC
+	component = ic.newComponent(common.SlugBIOS, devBMC.Vendor, devBMC.Model, devBMC.Serial, devBMC.ProductName,
+		&rivets.ComponentAttributes{
+			Description:  devBMC.Description,
+			ProductName:  devBMC.ProductName,
+			Oem:          devBMC.Oem,
+			Metadata:     devBMC.Metadata,
+			Capabilities: devBMC.Capabilities,
+		},
+		devBMC.Status,
+		devBMC.Firmware,
+	)
+	if component != nil {
+		components = append(components, component)
+	}
+
+	// mainboard
+	devMainBoard := device.Mainboard
+	component = ic.newComponent(common.SlugBIOS, devMainBoard.Vendor, devMainBoard.Model, devMainBoard.Serial, devMainBoard.ProductName,
+		&rivets.ComponentAttributes{
+			Description:  devMainBoard.Description,
+			ProductName:  devMainBoard.ProductName,
+			Oem:          devMainBoard.Oem,
+			PhysicalID:   devMainBoard.PhysicalID,
+			Metadata:     devMainBoard.Metadata,
+			Capabilities: devMainBoard.Capabilities,
+		},
+		devMainBoard.Status,
+		devMainBoard.Firmware,
+	)
+	if component != nil {
+		components = append(components, component)
+	}
+
+	// memory
```

Comment:
If I am converting the bios/bmc/mainboard correctly, I will continue fill memory/nics/drivers... component.

Reply:
This looks good so far.